src/persist-client/src/impl/machine.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Implementation of the persist state machine.
+
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use anyhow::anyhow;
+use differential_dataflow::difference::Semigroup;
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::trace::Description;
+use mz_persist::location::{Consensus, LocationError, SeqNo, VersionedData};
+use mz_persist_types::{Codec, Codec64};
+use timely::progress::{Antichain, Timestamp};
+use tracing::debug;
+
+use crate::error::{InvalidUsage, NoOp};
+use crate::r#impl::state::{ReadCapability, State, WriteCapability};
+use crate::read::ReaderId;
+use crate::write::WriterId;
+use crate::Id;
+
+#[derive(Debug)]
+pub struct Machine<K, V, T, D> {
```

Comment:
I really like how Machine, State, and StateMeta turned out. inc state also seems to fit pretty nicely into this framing